File: backend/routers/members.py
"""Team members CRUD for the web platform (manager-gated)."""
from __future__ import annotations
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from .. import access_control, members_store
from ..webapp_auth import COOKIE_NAME, dashboard_auth_enabled, session_role
router = APIRouter()
def _require_manager(request: Request) -> str:
    if not dashboard_auth_enabled():
        return "owner"
    role = session_role(request.cookies.get(COOKIE_NAME))
    if not access_control.is_manager(role):
        raise HTTPException(status_code=403, detail="Manager access required.")
    return role
class AddMemberRequest(BaseModel):
    username: str | None = None
    userId: str | None = None
    role: str
class SetRoleRequest(BaseModel):
    role: str
@router.get("/api/members")
def list_members(request: Request) -> dict:
    _require_manager(request)
    return {"members": members_store.list_members()}
@router.post("/api/members")
def add_member(body: AddMemberRequest, request: Request) -> dict:
    actor = _require_manager(request)
    try:
        member = members_store.add_member(user_id=body.userId, username=body.username, role=body.role, added_by=actor)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return {"ok": True, "member": member}
@router.patch("/api/members/{key:path}")
def set_role(key: str, body: SetRoleRequest, request: Request) -> dict:
    actor = _require_manager(request)
    try:
        member = members_store.set_role(key, body.role, actor=actor)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except KeyError as e:
        raise HTTPException(status_code=404, detail="Member not found.") from e
    return {"ok": True, "member": member}
@router.delete("/api/members/{key:path}")
def remove_member(key: str, request: Request) -> dict:
    actor = _require_manager(request)
    try:
        members_store.remove_member(key, actor=actor)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except KeyError as e:
        raise HTTPException(status_code=404, detail="Member not found.") from e
    return {"ok": True}

File: src/services/members.ts
export type MemberRole = 'owner' | 'admin' | 'viewer'

export interface Member {
  userId: string | null
  username: string | null
  role: MemberRole
  addedBy?: string
  addedAt?: string
}

export function memberKey(member: Member): string {
  return member.userId ?? 'u:' + member.username
}

export async function listMembers(): Promise<Member[]> {
  const response = await fetch('/api/members')
  if (!response.ok) {
    throw new Error('Not authorized to view the team.')
  }
  const json = (await response.json()) as { members?: Member[] }
  return json.members ?? []
}

export async function addMember(input: { username?: string; userId?: string; role: 'admin' | 'viewer' }): Promise<void> {
  const response = await fetch('/api/members', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
  if (!response.ok) {
    const json = (await response.json().catch(() => ({}))) as { detail?: string }
    throw new Error(json.detail ?? 'Could not add the member.')
  }
}

export async function setMemberRole(key: string, role: 'admin' | 'viewer'): Promise<void> {
  const response = await fetch(`/api/members/${encodeURIComponent(key)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ role }),
  })
  if (!response.ok) {
    const json = (await response.json().catch(() => ({}))) as { detail?: string }
    throw new Error(json.detail ?? 'Could not change the role.')
  }
}

export async function removeMember(key: string): Promise<void> {
  const response = await fetch(`/api/members/${encodeURIComponent(key)}`, { method: 'DELETE' })
  if (!response.ok) {
    const json = (await response.json().catch(() => ({}))) as { detail?: string }
    throw new Error(json.detail ?? 'Could not remove the member.')
  }
}
